### PR TITLE
feat(search): allow searching for external ids

### PR DESCRIPTION
**Pull request by smrutis1**
Signed-off-by: Smruti Prakash Sahoo <smruti.sahoo@siemens.com>

[//]: # (Copyright Bosch.IO GmbH 2020)
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

> Please provide a summary of your changes here.
> * Which issue is this pull request belonging to and how is it solving it? (*#1551*)
> * Did you add or update any new dependencies that are required for your change? - No

Issue: 

* try to search with examples given in the issues

> Have you implemented any additional tests? - No

### Checklist
Must:
- [ ] All related issues are referenced in commit messages and in PR

### CODE REVIEW

**Review by akapti — COMMENTED** [review]
Code looks good to me.
Testing Review: But the thing is I am not sure if it's possible to perform search on `@` character in Lucene, also every time there is a any special characters like `+ - = && || > < ! ( ) { } [ ] ^ " ~ * ? : \ / @` in search string, it sanitize the search string (breaks) into tokens and returns matching result with each token.

**Comment by rudra-superrr** [comment]
Testing Review:
**Issue in Restrict To Type section** i.e. while searching keyword for particular type(s), it showing result for other type(s) also.
For ex. searching keyword (external id) 'pkg:deb/debian/base-files@11.1+deb11u3?arch=source' for Release type only, then the result will show the keyword present in different project(s) as well as in other Types (Component, Releases etc) also
search
.

**Comment by smrutis1** [comment]
@akapti (from discussion) .  If the search string starts with `pkg:` it will do a exact search of the term.
@rudra-superrr , the mentioned issue is fixed, Kindly verify.

**Comment by rudra-superrr** [comment]
Issue in Restrict to Type section is fixed.

**Comment by rudra-superrr** [comment]
Restrict to type section issue is fixed.

File: backend/src/src-search/src/main/java/org/eclipse/sw360/search/SearchHandler.java [hunk 1]
```diff
@@ -62,19 +71,82 @@ public List<SearchResult> searchFiltered(String text, User user, List<String> ty
...
+        Collections.sort(srs, new SearchResultComparator());
 
         if (log.isTraceEnabled())
             log.trace("Search for " + text + " returned " + results.size() + " results");
 
-        return results;
+        return srs;
+    }
+
+    private void dealWithSpecialCharacters(String text, User user, Set<SearchResult> results, AbstractDatabaseSearchHandler dbSearchHandler, final List<String> typeMask) {
```
Comment (akapti) [inline backend/src/src-search/src/main/java/org/eclipse/sw360/search/SearchHandler.java:99]: Please re-use the code  y making another generic method.

**Review by akapti — CHANGES_REQUESTED** [review]
Suggested some review comments.

File: backend/src/src-search/src/main/java/org/eclipse/sw360/search/db/AbstractDatabaseSearchHandler.java [hunk 2]
```diff
@@ -136,4 +154,15 @@ private static SearchResult makeSearchResult(LuceneResult.Row row) {
 
         return result;
     }
+
+    private static String sanitizeQueryInput(String input) {
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply backend/src/src-search/src/main/java/org/eclipse/sw360/search/db/AbstractDatabaseSearchHandler.java]: removed

File: backend/src/src-search/src/main/java/org/eclipse/sw360/search/SearchHandler.java [hunk 3]
```diff
@@ -62,19 +71,82 @@ public List<SearchResult> searchFiltered(String text, User user, List<String> ty
         }
 
         // Query user and other database
-        List<SearchResult> results = Lists.newArrayList();
+        Set<SearchResult> results = Sets.newLinkedHashSet();
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply backend/src/src-search/src/main/java/org/eclipse/sw360/search/SearchHandler.java:74]: Removed

File: backend/src/src-search/src/main/java/org/eclipse/sw360/search/SearchHandler.java (hunk 1 above)
Comment (smrutis1) [reply backend/src/src-search/src/main/java/org/eclipse/sw360/search/SearchHandler.java:99]: Done

**Comment by smrutis1** [comment]
@akapti Have incorporated the changes. Please have a look.

**Review by akapti — CHANGES_REQUESTED** [review]
There are compilation errors on line 103 and 105 in this file `backend/src/src-search/src/main/java/org/eclipse/sw360/search/SearchHandler.java`.
Please fix it.

**Comment by smrutis1** [comment]
Oops :( . Somehow I missed it after refactoring. Corrected it :(

**Review by akapti — CHANGES_REQUESTED** [review]
Tested the changes, working as expected.
Only one minor review comment to remove the unused imports

**Review by akapti — APPROVED** [review]
Code looks good.

**Pull request merged**